File: tps2/tp1/Palindromo.java
/**
 *
 * Pontificia Universidade Catolica de Minas Gerais
 *
 * Curso de Ciencia da Computacao
 * Algoritimo e Estruturas de Dados II
 *
 * TP01Q01 - Palíndromo - v1.0 - 28 / 08 / 2025
 * 855796 - Rubens Dias Bicalho
 *
 */

import java.util.Scanner;

public class Palindromo {

    // Verifica se duas strings são iguais
    public static boolean equals(String a, String b) {
        int a_len = a.length();
        int b_len = b.length();
        if (a_len != b_len) {
            return false;
        }

        for (int i = 0; i < a_len; i++) {
            if (a.charAt(i) != b.charAt(i)) {
                return false;
            }
        }
        return true;
    }

    // Verifica se a string passada é palíndromo
    public static boolean isPalindromo(String input) {
        int length = input.length();
        for (int i = 0; i < length / 2; i++) {
            if (input.charAt(i) != input.charAt(length - i - 1)) {
                return false;
            }
        }
        return true;
    }

    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);
        String input = sc.nextLine();
        while (!(equals(input, "FIM"))) {
            System.out.println(isPalindromo(input) ? "SIM" : "NAO");
            input = sc.nextLine();
        }
        sc.close();
    }

}

File: tps2/tp1/palindromo.c
/**
 *
 * Pontificia Universidade Catolica de Minas Gerais
 *
 * Curso de Ciencia da Computacao
 * Algoritimo e Estruturas de Dados II
 *
 * TP01Q02 - Palíndromo RECURSIVO - v1.0 - 28 / 08 / 2025
 * 855796 - Rubens Dias Bicalho
 *
 */

#include <locale.h>
#include <stdio.h>
#include <string.h>
#include <stdbool.h>
#include <wchar.h>
#include <wctype.h>

// Checa se as duas strings são palindromos de forma recursiva
bool isPalindromo(wchar_t input[], int start, int end) {
    if (start >= end) return true;
    if (input[start] != input[end]) return false;

    return isPalindromo(input, start + 1, end - 1);
}

int main() {

    setlocale(LC_ALL, "");

    wchar_t message[100]; // Usa widechar por causa do charset dos arquivos de entrada


    while (1) {
        wint_t input;
        int length = 0;

        while ((input = getwchar()) != '\n') { // Lê caracteres até a nova linha
            if (length < 100) {
                message[length++] = input;
            }
        }
        message[length] = L'\0';

        if (wcscmp(message, L"FIM") == 0) {
            break;
        }

        isPalindromo(message, 0, length - 1) ? wprintf(L"SIM\n") : wprintf(L"NAO\n");
    }

    return 0;
}
    